Sources/Hyperspace/Service/Network/Cancellation.swift

```diff
@@ -0,0 +1,78 @@
+//
+//  Cancellation.swift
+//  Hyperspace
+//
+//  Created by Will McGinty on 2/19/19.
+//  Copyright © 2019 Bottle Rocket Studios. All rights reserved.
+//
+
+import Foundation
+
+/// Manages cancellation tokens and signals them when cancellation is requested.
+///
+/// All `CancellationTokenSource` methods are thread safe.
+public final class CancellationSource {
+    
+    public struct Token {
+        // MARK: Properties
+        private weak var source: CancellationSource?
+        public var isCancelling: Bool { return source?.isCancelling ?? false }
+        
+        public init(source: CancellationSource?) {
+            self.source = source
+        }
+        
+        // MARK: Interface
+        public func register(closure: @escaping () -> Void) {
+            source?.register(closure)
+        }
+    }
+    
+    // MARK: Properties
+    private let lock = NSLock()
+    private var observers: [() -> Void]? = []
+    
+    // MARK: Initializers
+    public init() {}
+    
+    // MARK: Interface
+    public var isCancelling: Bool {
+        lock.lock()
+        defer { lock.unlock() }
+        return observers == nil
+    }
+    
+    public var token: CancellationSource.Token { return CancellationSource.Token(source: self) }
```

Comment:
Isn't `CancellationSource` suppose to manage cancellation tokens? It only appears to keep track of one token.

Reply:
It keeps track of as many Token's as you create using it as a source. If you check out the `register` function of the `Token`, it calls into `CancellationSource.lockedRegister` which registers the closure as an observer (inserting into it's Array storage). As soon as you call `cancel` on the `CancellationSource` it runs it's entire array of observers to perform the cancellation.

Reply:
So let me see if I understand this correctly:
I can create a `CancellationSource` and give it a `CancellationSource.Token`. I can register closures with the token which registers the closures in the source. When I ask the token for `isCancelling`, it asks the source for `isCancelling`. When I call `cancel()` on the source, it runs through the array of observers to perform the closures without concern for the token.
I am failing to see the need of having a token when all it does is forward or get things from the source.

Reply:
You don't necessarily give the `CancellationSource` a `Token`. The source creates tokens on its own using the fileprivate initializer. For you, as the end user, to get a token, you just ask the source.

Let's say you have multiple concurrent tasks, or even a single task involving a series of serial subtasks. In most cases, you'll want to have 1 single source of cancellation. So in this case, if you tell the source to cancel, it will cancel each of its tokens automatically meaning all subtasks will be cancelled. If you only have 1 token active from a source, only 1 will be cancelled, but the API to do so remains the same whether your cancelling 1 or 20 tasks.

Reply:
That makes a bit more sense. So if I had several tokens and wanted to cancel just one of them, how would I do that? or is that beyond the scope of what this is intended to accomplish by being able to cancel all requests

Reply:
We talked briefly about this and came up with some solutions

1) You could create your tokens each with different sources.
2) We could add additional identifying information (perhaps a UUID) to each token so that they can be differentiated from one another within a single cancellation source.

Reply:
My understanding is much improved. Thanks @wmcginty